lib/actions/ssh-direct.ts

```diff
@@ -0,0 +1,245 @@
+		};
+		const proxyData = (chunk: Buffer) => {
+			if (chunk != null) {
+				tunnelProxyResponse += chunk.toString();
+			}
+
+			// read 'data' chunks until full HTTP status line has been read
+			if (!_.includes(tunnelProxyResponse, '\r\n\r\n')) {
+				return;
+			}
+			socket.removeListener('data', proxyData);
+			socket.removeListener('end', earlyEnd);
+			socket.removeListener('error', earlyError);
+
+			// RFC2616: Status-Line = HTTP-Version SP Status-Code SP Reason-Phrase CRLF
+			let httpStatusLine = tunnelProxyResponse.split('\r\n')[0];
+			const httpStatusCode = parseInt(httpStatusLine.split(' ')[1], 10);
+
+			// translate "Proxy-authorization required" to more user-friendly "Forbidden"
+			if (httpStatusCode === 407) {
+				httpStatusLine = 'HTTP/1.0 403 Forbidden';
+			}
+
+			if (httpStatusCode !== 200) {
+				console.error(
+					`Could not connect to ${host}:${port} - ${httpStatusLine}`,
+				);
+				return reject(
+					new Error(`Could not connect to ${host}:${port} - ${httpStatusLine}`),
+				);
+			}
+
+			// one proxied socket, ready to go!
+			resolve(socket);
+		};
+
+		socket
+			.on('end', earlyEnd)
+			.on('error', earlyError)
+			.on('data', proxyData);
+	});
+
+Bluebird.promisifyAll(ssh2.Client);
+
+interface Ssh2ClientAsync extends ssh2.Client {
+	execAsync(
+		cmd: string,
+		options?: ssh2.ExecOptions,
+	): Bluebird<ssh2.ClientChannel>;
+}
+
+const createSSHClient = (socket: stream.Duplex): Bluebird<Ssh2ClientAsync> =>
+	new Bluebird((resolve, reject) => {
+		const client = new ssh2.Client() as Ssh2ClientAsync;
+		return client
+			.on('ready', () => {
+				resolve(client);
+			})
+			.on('error', err => {
+				const errSource = err && err.level ? 'ssh client socket' : 'ssh client';
+				let errMsg = `${errSource} error while initiating SSH connection`;
+				if (err && err.description) {
+					errMsg += `: ${err.description}`;
+				}
+				console.error(errMsg);
+				reject(new Error(errMsg));
+			})
+			.connect({
+				sock: socket,
+				username: 'root',
+				port: 22222,
+				agent: process.env.SSH_AUTH_SOCK,
+			});
+	});
+
+const createSSHClientDisposer = (
+	socket: stream.Duplex,
+): Bluebird.Disposer<Ssh2ClientAsync> =>
+	createSSHClient(socket).disposer((client: Ssh2ClientAsync) => client.end());
+
+interface ExecResponse {
+	code: number;
+	data?: {
+		stdout: string;
+		stderr: string;
+	};
+	signal?: string;
+}
+
+const pipeDeviceToStdout = (
+	deviceStream: ssh2.ClientChannel,
+): Bluebird<ExecResponse> => {
+	const dsResize = () =>
+		deviceStream.setWindow(process.stdout.rows!, process.stdout.columns!, 0, 0);
+	const dsEnd = () => deviceStream.end();
+	const dsWrite = (data: Buffer) => deviceStream.write(data);
+
+	if (process.stdin != null && process.stdout.isTTY) {
+		if (typeof process.stdin.setRawMode === 'function') {
+			process.stdin.setRawMode(true);
+		}
+
+		// Set initial window size and handle resize events
+		dsResize();
+		process.stdout.on('resize', dsResize);
```

Comment:
So we need to deal with window resizing in our code... :-/ Not sure if a blessing or a curse:
[Broken line wrapping / cursor behavior with balena ssh](https://github.com/balena-io/balena-cli/blob/master/TROUBLESHOOTING.md#broken-line-wrapping--cursor-behavior-with-balena-ssh)

I guess a new troubleshooting item will be: _"try the `-x` option"..._ :-)

Reply:
Oh - "today I learned" :-) that the CLI already had some code to listen to the stdout 'resize' event, though probably not executed for the `balena ssh` command:
https://github.com/balena-io/balena-cli/blob/v11.25.12/lib/utils/tty.coffee
It is currently being converted to Typescript in PR #1602 (becoming `tty.ts`).
Maybe there's some room for code reuse, though not necessary -- I simply thought I would mention it.